opencog/reasoning/RuleEngine/rule-engine-src/pln/DefaultForwardChainerCB.cc

```diff
@@ -25,111 +25,186 @@
-				bindlinks.push_back(hb);
-		}
-	}
-	//transfer back bindlinks to main atomspace and  get their handle
-	HandleSeq copied_back;
-	for (Handle h : bindlinks)
-		copied_back.push_back(as_->addAtom(h));
-	//find the rules containing the bindLink in copied_back
-	vector<Rule*> matched_rules;
-	for (Rule* r : rules)
-		if (find(copied_back.begin(), copied_back.end(), r->get_handle())
-				!= copied_back.end())
-			matched_rules.push_back(r);
-	return matched_rules;
+    Handle target = fcmem.get_cur_target();
+    if (target == Handle::UNDEFINED or NodeCast(target))
+        throw InvalidParamException(TRACE_INFO,
+                                    "Needs a target atom of type LINK");
+    HandleSeq chosen_bindlinks;
+
+    if (LinkCast(target)) {
+        AtomSpace rule_atomspace;
+        SchemeEval* sceval = new SchemeEval(&rule_atomspace);
+        //Handle target_cpy=rule_atomspace.addAtom(target); xxx this doesn't work
+        Handle target_cpy = sceval->eval_h(SchemeSmob::to_string(target)); //xxx this works
+
+        //copy rules to the temporary atomspace
+        vector<Rule*> rules = fcmem.get_rules();
+        for (Rule* r : rules) {
+            //rule_atomspace.addAtom(r->get_handle()); xxx this doesn't work
+            sceval->eval_h(SchemeSmob::to_string(r->get_handle())); //xxx this works
+        }
+
+        //create bindlink with target as an implicant
+        PLNCommons pc(&rule_atomspace);
+        Handle copy = pc.replace_nodes_with_varnode(target_cpy, NODE);
+        Handle bind_link = pc.create_bindLink(copy, false);
+
+        //pattern match
+        DefaultImplicator imp(&rule_atomspace);
+        try {
+            PatternMatch pm;
+            pm.do_bindlink(bind_link, imp);
+        } catch (InvalidParamException& e) {
+            cout << "VALIDATION FAILED:" << endl << e.what() << endl;
+        }
+
+        //get matched bindLinks
+        HandleSeq matches = imp.result_list;
+        if (matches.empty()) {
+            logger().debug(
+                    "No matching BindLink was found.Returning empty vector");
+            return vector<Rule*> { };
+        }
+
+        HandleSeq bindlinks;
+        for (Handle hm : matches) {
+            //get all BindLinks whose part of their premise matches with hm
+            HandleSeq hs = get_rootlinks(hm, &rule_atomspace, BIND_LINK);
+            for (Handle hi : hs) {
+                if (find(bindlinks.begin(), bindlinks.end(), hi) == bindlinks.end()) {
+                    bindlinks.push_back(hi);
+                }
+            }
+        }
+        delete sceval; //delete to use SchemeEval with another atomspace.this might be a hack.
+
+        //push back handles to main atomspace
+        for (Handle h : bindlinks) {
+            //auto bindlink = as_->addAtom(h);
+            SchemeEval *sc = new SchemeEval(as_);
+            auto bindlink = sc->eval_h(SchemeSmob::to_string(h));
+            chosen_bindlinks.push_back(bindlink);
```

Comment:
Yep.  Found it it is fixed in fc31a4b24e6a3b35dacebaf7865c90c261283581  and a unit test tests in in 11ef7f927d8f14ebb34cb4f8d8913f0ff9de1342